### ISSUE TITLE
Wifi infrared blaster

### ISSUE THREAD

**Issue by Zuikkis** [codetheweb/tuyapi#134]
Howdy!

I bought this cheap wifi IR controller:

https://www.aliexpress.com/item/Mini-Smart-Switch-WIFI-Wireless-Remote-Control-Smart-Home-Automation-For-TV-Air-Conditioner-IR-Infrared/32961659020.html

I planned to use it in Domoticz home automation, but I have installed tuyapi/cli for testing.

I used the "recommended method" to get devid and key. I can use tuya-cli set and get with these values and it talks to the device correctly. However, there seems to be only one DPS called "send_ir". Obviously that would need some kind of additional parameter to indicate what IR to send? 

In Tuya Smart android app you can configure the device to control several different appliances like TV, DVD, air conditioner etc. My main interest was the aircon.

I used some Android network capture tool to see what the app is sending (to the cloud) when I use the app to send IR codes. I got fairly good looking JSON:

```
Request Head:

POST /api.json? .... 
(clipped)
...
Content-Type: application/x-www-form-urlencoded
Content-Length: 274
Host: a1.tuyaeu.com
Connection: Keep-Alive
Accept-Encoding: gzip


Body:

postData={"devId":"*deleted*","dps":"{\"101\":true,\"102\":\"1\",\"103\":21,\"104\":\"2\"}","dpsTime":"1548957769","gwId":"*deleted*}

Response Head:

HTTP/1.1 200 OK
Date: Thu, 31 Jan 2019 18:02:49 GMT
Content-Type: application/json;charset=UTF-8
Content-Length: 63
Connection: keep-alive
Server: Apache-Coyote/1.1
Content-Language: zh-CN


Body:

{"result":true,"t":1548957769244,"success":true,"status":"ok"}

```

I deleted gwid and devid so you don't steal my devices. :) But it's worth noting that those are _different_ values. Gwid is the id that was shown during installation, devid is something else.

The "devid" seems to have a new set of DPS which were quite easy to find out by clicking all the buttons.. It's for the aircon:

  101    true= power on, false= power off
  102    mode, 0= cool, 1=heat, 2=auto, 3=fan, 4=dry
  103    temp C
  104    0-3,  fan setting (0=auto)

So there's like two levels of device id's here. Is this supported in the api? 

I assume that if I add codes for another device like a TV, it will create a new "devid" but "gwid" still is the same. Didn't try it though.

**Comment by Zuikkis** [comment]
Oh, I should have mentioned I already tried such approach. :) tuya-cli just says "Set succeeded" but nothing happens in the air conditioner.  Response is the same with any --id, btw. Even if I just type random letters. 

By the way, the devid picked up from the JSON looks a bit weird because it's not entirely hexadecimal. Is this normal? It first has 18 hexadecimal digits and then "cjbi" in the end..

**Comment by codetheweb** [comment]
Huh, that doesn't sound normal...

Are you able to sniff the control packets the app sends with Wireshark and send the .pcap to me?

**Comment by codetheweb** [comment]
Interesting...

The reason I was asking for a packet capture is that the syntax the app uses when communicating with the cloud is typically slightly different than when it communicates with a device locally.  

Please do try a short program when you have time and let me know what happens. :)

**Comment by Zuikkis** [comment]
Hmm. I didn't see any local access between "Tuya Smart" app and the device. App always just sends to the cloud.


I wrote a small program, but it didn't work. :(

Here's my code:

```
const TuyAPI = require('tuyapi');

const device = new TuyAPI({
  id: 'xxxx',
  key: 'yyyyy',
  ip: 'nnn.nnn.nnn.nnn'});

device.set({
    multiple: true,
    data: {
      '101': true,
      '102': '1',
      '103': 24,
      '104': '3'
    }}) .then(result => {
       console.log('Result: ' + result);

});

```

It "works" in that it says "Result: true". So the command is accepted. But it does not control my aircon. :(

If I modify the key, it no longer succeeds. So key is correct. However id can be anything and it always returns "true". Perhaps the non-hexadecimal id is not passed correctly?

**Comment by codetheweb** [comment]
Try running your script with a prefix of `DEBUG=* ` (so `DEBUG=* node yourscript.js`) and post the output.

If the app only communicates with the cloud to control your device that could be an issue.  Based solely on the noticeable latency, do you think that the app always uses the cloud API?

**Comment by Zuikkis** [comment]
```
  TuyAPI Payload: 192.168.0.133 +0ms
  TuyAPI { devId: 'xxxxxxxxxxxxxxxx',
  TuyAPI   uid: '',
  TuyAPI   t: '1549345280',
  TuyAPI   dps: { '101': true, '102': '1', '103': 24, '104': '3' } } +15ms
  TuyAPI Send attempt 1 +56ms
  TuyAPI Sending this data: 000055aa0000000000000007000000b3332e3162376533373464323631343461393932596467734b4d4c2b51592b3550354b6c35686b2f616e55543659505575617575442f652b72533676486f4134645968644b386631763776714932706c45552b5a386c6136543070536136634e6c564c4e792b496d3666357431576c664f7a3958796c4d4a6a4d7a3678714d42326b39696b725278646a6d6d58337869315447784e6d356845722f584e4c5a2b6f5254344b48386b71673d3d000000000000aa55 +2ms
  TuyAPI Connect 192.168.0.133 +5ms
  TuyAPI Socket connected. +51ms
  TuyAPI Received data back: 192.168.0.133 +18ms
  TuyAPI 000055aa00000000000000070000000c00000000789370910000aa55 +1ms
  TuyAPI undefined 192.168.0.133 undefined 7 +5ms
  TuyAPI Disconnect +1ms
Result: true
  TuyAPI Socket closed: 192.168.0.133 +11ms

```

The json looks ok? Only difference is single quotes instead of double as in my original post, but shouldn't make a difference..

**Comment by Zuikkis** [comment]
Oh, and original data has separate devId and gwId, with different values. Is there a way to add gwId to json generated by tuyapi?

**Comment by codetheweb** [comment]
Just published a new version (v3.2.3) that adds an option for a gateway ID.  Take a look at the [docs](https://codetheweb.github.io/tuyapi/#tuyadevice) for details.

Let me know if that works.

**Comment by Zuikkis** [comment]
Thanks for you efforts! Now json looks correct, but it still doesn't work.

Maybe this device does not support local access at all? I have never seen local access from the android app when I have logged it. Always talks to the cloud only.

**Comment by codetheweb** [comment]
It's possible, although it seems unlikely...

You may want to attempt to flash new firmware on it, as controlling it through the cloud requires API keys that you have to pay for.  Take a look at [this](https://github.com/ct-Open-Source/tuya-convert) project.

**Comment by codetheweb** [comment]
Great, glad you were able to get it working. :)

**Comment by mlt** [comment]
@Zuikkis Would you mind sharing some insights? Is it possible to back up original firmware before flashing? Are you using Google Home by chance to control AC? If so, how did you set it up without Tuya cloud? IFTTT? But don't you then need to "Talk to &hellip;"? I just got a similar IR blaster but I'm already disappointed with default software. All I can do is turn TV on/off whereas it would be nice to be able to send a key sequence bound to a certain voice command with Google Home. I can see *Scenes* in *Smart Life* app, but there is no way to set up details for *smart IR:send_ir* command.

**Comment by Zuikkis** [comment]
Yeah, you can find the necessary info on Sonoff-Tasmota site. There is a dedicated page for this IR blaster:

https://github.com/arendst/Sonoff-Tasmota/wiki/YTF-IR-Bridge

My device looked slightly different but works exactly the same way.

I flashed mine directly with ftdi cable. It's easy to do a backup of the original firmware. 

You need at least version 6.4.1.8 of Tasmota. The latest stable is not new enough, but there are pre-built development versions that are newer. You might want to compile it yourself though, to set the compile time options..

**Comment by Pedro-vk** [comment]
It seems that there is no good way to get the specifics DPS to send the correct config to the IR device. I was able to intercept the app request but the API changed, and now it is sending the data encrypted, so one more layer to this complicated and weird thing.
